File: trees/inorder_traversal_of_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
    result = []

    def inorder(node, result)
        return if node.nil?

        inorder(node.left, result)
        result << node.val
        inorder(node.right, result)
    end
    inorder(root, result)
    result
end

File: trees/inorder_traversal_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[int]
        """

        result = []
        
        def inorder(node):
            if root is None:
                return

            self.inorder(node.left):
            result.append(node.val)

            print(root.val, end=' ')

            self.inorder(root.right)
        self.inorder(root)
        return result